### PR TITLE
Learning rates based on Karpathy CS231n lecture 6

### PR DESCRIPTION
**Pull request by chemelnucfin**
Continuation of #1512 
@vrv seems like I can test now.  Updated the pull request as you suggested.  Please let me know if you have any comments.

### CODE REVIEW

File: tensorflow/python/training/learning_rate_decay.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from __future__ import print_function
 
 from tensorflow.python.framework import ops
-from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import math_ops, constant_op
```
Comment (vrv) [inline tensorflow/python/training/learning_rate_decay.py]: please import one module per line

File: tensorflow/python/training/learning_rate_decay_test.py [hunk 2]
```diff
@@ -73,5 +75,67 @@ def testVariables(self):
...
 
+class ExponentialDecayTest(test_util.TensorFlowTestCase):
+
+  def testDecay(self):
+    initial_lr = 0.1
+    k = 10
+    decay_rate = 0.96
+    with self.test_session():
+      for step in range(0, k+1):
+        lr = learning_rate_decay.natural_exp_decay(
```
Comment (vrv) [inline tensorflow/python/training/learning_rate_decay_test.py]: these tests are creating 10 copies of the graph -- can you rewrite the test so that it creates the graph once and passes the proper parameters as inputs to the function? 


You can see how it's done for the exponential case above: all the ops are created first rather than creating them in a loop.

**Comment by rwill128** [comment]
I find it helpful to phrase commit titles in the imperative, with a verb that describes what merging that commit will do. Basically I try to write commit messages that can complete this sentence: Adding this commit will <do something useful>. Using this commit as an example, it would be helpful for me to know what's being done with the learning rates -- is a new library method being added, or is it becoming the new default learning rate, etc.

I'm a new guy and I submit this as a humble comment.

File: tensorflow/python/training/learning_rate_decay_test.py (hunk 2 above)
Comment (chemelnucfin) [inline tensorflow/python/training/learning_rate_decay_test.py]: @vrv is the way I wrote it now better?  Thanks.

File: tensorflow/python/training/learning_rate_decay_test.py [hunk 3]
```diff
@@ -108,5 +110,75 @@ def testPiecewiseConstantEdgeCases(self):
...
+class ExponentialDecayTest(test_util.TensorFlowTestCase):
+
+  def testDecay(self):
+    initial_lr = 0.1
+    k = 10
+    decay_rate = 0.96
+    with self.test_session():
+      step = state_ops.variable_op([], dtypes.int32)
+      for i in range(k+1):
+        assign = state_ops.assign(step, i)
```
Comment (vrv) [inline tensorflow/python/training/learning_rate_decay_test.py]: to make it clear, it is best practice to create the entire graph ahead of time and then invoke it by calling .eval() or sess.run().

This is adding a new op (assign) for every iteration of the loop.

What about moving this statement out of the loop and changing it to assign_add(step, 1) instead ?

**Comment by chemelnucfin** [comment]
@vrv Is this what you mean?

File: tensorflow/python/training/learning_rate_decay_test.py [hunk 4]
```diff
@@ -108,5 +110,82 @@ def testPiecewiseConstantEdgeCases(self):
...
+    initial_lr = 0.1
+    k = 10
+    decay_rate = 0.96
+    step = state_ops.variable_op([], dtypes.int32)
+    assign = state_ops.assign(step, 0)
+    add = state_ops.assign_add(step, 1)
+    with self.test_session():
+      assign.op.run()
+      for i in range(k+1):
+        decayed_lr = learning_rate_decay.natural_exp_decay(initial_lr, step,
```
Comment (vrv) [inline tensorflow/python/training/learning_rate_decay_test.py]: This calls natural_exp_delay once for every value of 'i', which means you are still adding 'k' versions of the ops in this loop.

The idea is to define the entire graph completely, before even with self.test_session(), so that the code within self.test_session() is just driving the existing graph, rather than modifying it.  Does that make sense?

File: tensorflow/python/training/learning_rate_decay_test.py [hunk 5]
```diff
@@ -215,5 +217,83 @@ def testBeyondEndWithCycle(self):
...
+    decay_rate = 0.96
+    step = state_ops.variable_op([], dtypes.int32)
+    assign_step = state_ops.assign(step, 0)
+    increment_step = state_ops.assign_add(step, 1)
+    decayed_lr = learning_rate_decay.natural_exp_decay(initial_lr,
+                                                       step,
+                                                       k,
+                                                       decay_rate,
+                                                       staircase=True)
+    with self.test_session():
```
Comment (chemelnucfin) [inline tensorflow/python/training/learning_rate_decay_test.py]: @vrv I submitted a new pull request, is that what you mean?  Thanks a lot!

Comment (vrv) [inline tensorflow/python/training/learning_rate_decay_test.py]: Hm, I don't see a new pull request! :(

Comment (chemelnucfin) [inline tensorflow/python/training/learning_rate_decay_test.py]: @vrv oops, i meant as a force push...

Is this what you mean by moving the decay rate outside?

Thanks!

File: tensorflow/python/training/learning_rate_decay_test.py (hunk 4 above)
Comment (vrv) [inline tensorflow/python/training/learning_rate_decay_test.py]: I still see this being called within the range loop, am I not seeing the latest version?

File: tensorflow/python/training/learning_rate_decay_test.py (hunk 5 above)
Comment (chemelnucfin) [inline tensorflow/python/training/learning_rate_decay_test.py]: @vrv let me check the code again..   I think i didn't change it for all of
the tests, but only for test_staircase..  I'll get back to you when i am
back on my laptop.

Comment (chemelnucfin) [inline tensorflow/python/training/learning_rate_decay_test.py]: @vrv apolgies for the confusion.

Comment (chemelnucfin) [inline tensorflow/python/training/learning_rate_decay_test.py]: @vrv please try this and let me know if there are any problems!  Thanks for your patience in dealing with this PR.

**Comment by vrv** [comment]
Great!

@tensorflow-jenkins test this please
